### PR TITLE
Updates to spoke, route table and nsg modules

### PR DESCRIPTION
**Pull request by jk-ans**
# Description

Updating the spoke module to use the route-table module and to use the subnet association feature in the nsg module

 - Use route-table module
 - Use nsg module subnet association
 - Update validataion in subnet and route table modules
 - Small changes to route table module
 - Add Advanced Example to spoke module
 - Update version constraints and provider features on spoke, route table and nsg modules
 - Add Flow Log and Flow Log resources options

## Type of change

Please select the option that is relevant.

  - [ ] Bug fix (non-breaking change which fixes an issue)
  - [*] New feature (non-breaking change which adds functionality)
  - [*] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [*] This change requires a documentation update

# Checklist:

  - [*] My code follows the style guidelines of this project
  - [*] I have performed a self-review of my code
  - [*] I have commented my code, particularly in hard-to-understand areas
  - [*] I have made corresponding changes to the documentation
  - [*] My changes generate no new warnings e.g. code analysis tooling or general use of the changed code

### CODE REVIEW

**Comment by djackson-ans** [comment]
Trying to run the advanced example, I got this message:

```
Original Error: autorest/azure: Service returned an error. Status=<nil> Code="StorageAccountAlreadyTaken" Message="The storage account named flsatfmexadvspoke is already taken."
```

Perhaps the example should contain a random number / string generator for the storage account name?

**Comment by jk-ans** [comment]
> Trying to run the advanced example on the spoke, I got this message:
> 
> ```
> Original Error: autorest/azure: Service returned an error. Status=<nil> Code="StorageAccountAlreadyTaken" Message="The storage account named flsatfmexadvspoke is already taken."
> ```
> 
> Perhaps the example should contain a random number / string generator for the storage account name?

Added random_integer to storage account name

File: azure/spoke/variables.tf [hunk 1]
```diff
@@ -33,14 +33,19 @@ variable "address_space" {
 
   validation {
     error_message = "Must be valid IPv4 CIDR."
```
Comment (djackson-ans) [inline azure/spoke/variables.tf:35]: `Must be valid IPv4 CIDR.`

It's technically a list of valid CIDR's. Perhaps better phrasing would be:

`All Address Spaces must be Valid IPv4 CIDR`

File: azure/spoke/variables.tf [hunk 2]
```diff
@@ -33,14 +33,19 @@ variable "address_space" {
...
   }
 }
 
 variable "dns_servers" {
   description = "The DNS servers to use with this virtual network."
   type        = list(string)
   default     = []
+
+  validation {
+    error_message = "Value must be a valid IPv4 address."
```
Comment (djackson-ans) [inline azure/spoke/variables.tf:46]: As it's a list, phrasing would be:

`Values must be a valid IPv4 addresses.`

File: azure/network-security-group/examples/advanced/main.tf [hunk 3]
```diff
@@ -36,7 +53,7 @@ resource "azurerm_log_analytics_workspace" "nsg" {
 }
 
 resource "azurerm_storage_account" "nsg" {
-  name                = lower(replace("${local.resource_prefix}sa", "/[-_]/", ""))
+  name                = lower(replace("${local.resource_prefix}sa1", "/[-_]/", ""))
```
Comment (djackson-ans) [inline azure/network-security-group/examples/advanced/main.tf:56]: Perhaps randomize this name?

File: azure/spoke/examples/advanced/main.tf
Comment (djackson-ans) [inline azure/spoke/examples/advanced/main.tf:1]: As we're using the `random` provider, we need to make sure that we've added it to the example's `required_providers`:

EG:
```
terraform {
  required_providers {
    random = {
      source = "hashicorp/random"
      version = "3.6.0"
    }
  }
}

provider "random" {}
```

File: azure/network-security-group/examples/advanced/main.tf (hunk 3 above)
Comment (jk-ans) [reply azure/network-security-group/examples/advanced/main.tf:56]: added random number

**Review by djackson-ans — APPROVED** [review]

**Pull request merged**